File: apps/zxtune-android/zxtune/src/androidTest/java/app/zxtune/fs/VfsArchiveTest.java
package app.zxtune.fs;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import android.content.res.Resources;
import android.net.Uri;

import androidx.annotation.RawRes;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import app.zxtune.io.Io;
import app.zxtune.io.TransactionalOutputStream;
import app.zxtune.utils.StubProgressCallback;

@RunWith(AndroidJUnit4.class)
public class VfsArchiveTest {

  private Resources resources;
  private File tmpDir;

  private Uri getFile(@RawRes int res, String filename) throws IOException {
    final File result = new File(tmpDir, filename);
    final InputStream in = new BufferedInputStream(resources.openRawResource(res));
    final OutputStream out = new TransactionalOutputStream(result);
    final long size = Io.copy(in, out);
    assertTrue(size > 0);
    out.flush();
    out.close();
    return Uri.fromFile(result);
  }

  @Before
  public void setUp() {
    resources = InstrumentationRegistry.getInstrumentation().getContext().getResources();
    tmpDir = new File(System.getProperty("java.io.tmpdir", "."), String.format("VfsArchiveTest/%d",
        System.currentTimeMillis()));

  }

  @After
  public void tearDown() {
    final File[] files = tmpDir.listFiles();
    if (files != null) {
      for (File f : files) {
        f.delete();
      }
    }
    tmpDir.delete();
  }

  private static void unreachable() {
    fail("Unreachable");
  }

  @Test
  public void testNull() throws IOException {
    try {
      VfsArchive.resolve(null);
      unreachable();
    } catch (NullPointerException e) {
      assertNotNull("Thrown exception", e);
    }
    try {
      VfsArchive.browse(null);
      unreachable();
    } catch (NullPointerException e) {
      assertNotNull("Thrown exception", e);
    }
    try {
      VfsArchive.browseCached(null);
      unreachable();
    } catch (NullPointerException e) {
      assertNotNull("Thrown exception", e);
    }
  }

  @Test
  public void testDir() throws IOException {
    final Uri uri = Uri.parse("file:/proc");
    final VfsDir dir = (VfsDir) VfsArchive.resolve(uri);
    assertNotNull(dir);
    assertFalse(VfsArchive.checkIfArchive(dir));
    assertEquals("proc", dir.getName());
    assertEquals("", dir.getDescription());
    assertEquals(uri, dir.getUri());
  }

  @Test
  public void testFile() throws IOException {
    final Uri uri = Uri.parse("file:/proc/meminfo");
    final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
    assertNotNull(file);
    assertEquals("meminfo", file.getName());
    assertEquals("", file.getDescription());
    assertEquals("0.00B", file.getSize());
    assertEquals(uri, file.getUri());
  }

  @Test
  public void testFileTrack() throws IOException {
    final Uri uri = getFile(app.zxtune.test.R.raw.track, "track");
    final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
    assertNotNull(file);
    assertEquals("track", file.getName());
    assertEquals("", file.getDescription());
    assertEquals("2.8KB", file.getSize());

    final Uri normalizedUri = file.getUri();
    {
      final VfsFile browsed = (VfsFile) VfsArchive.browse(file);
      assertEquals(file, browsed);
    }
    {
      final VfsFile browsedCached = (VfsFile) VfsArchive.browseCached(file);
      assertEquals(file, browsedCached);
    }
    {
      final VfsFile resolved = (VfsFile) VfsArchive.resolve(uri);
      assertNotNull(resolved);
      assertEquals(normalizedUri, resolved.getUri());
    }
  }

  @Test
  public void testGzippedTrack() throws IOException {
    final Uri uri = getFile(app.zxtune.test.R.raw.gzipped, "gzipped");
    final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
    assertNotNull(file);
    assertEquals("gzipped", file.getName());
    assertEquals("", file.getDescription());
    assertEquals("54KB", file.getSize());

    final Uri normalizedUri = file.getUri();
    final Uri subUri = normalizedUri.buildUpon().fragment("+unGZIP").build();
    try {
      VfsArchive.resolve(subUri);
      unreachable();
    } catch (Exception e) {
      assertNotNull("Thrown exception", e);
      assertEquals("No archive found", e.getMessage());
    }
    {
      final VfsFile browsed = (VfsFile) VfsArchive.browse(file);
      assertEquals(file, browsed);
    }
    {
      final VfsFile browsedCached = (VfsFile) VfsArchive.browseCached(file);
      assertEquals(file, browsedCached);
    }
    {
      final VfsFile resolved = (VfsFile) VfsArchive.resolve(uri);
      assertNotNull(resolved);
      assertEquals(normalizedUri, resolved.getUri());
    }
    {
      final VfsFile resolvedFile = (VfsFile) VfsArchive.resolve(subUri);
      assertNotNull(resolvedFile);
      assertEquals(subUri, resolvedFile.getUri());
      assertEquals("+unGZIP", resolvedFile.getName());
      assertEquals("sll3", resolvedFile.getDescription());
      assertEquals("2:33", resolvedFile.getSize());

      // no parent dir really, so just file itself
      final VfsFile resolvedFileParent = (VfsFile) resolvedFile.getParent();
      assertNotNull(resolvedFileParent);
      assertEquals(normalizedUri, resolvedFileParent.getUri());
    }
  }

  @Test
  public void testMultitrack() throws IOException {
    final Uri uri = getFile(app.zxtune.test.R.raw.multitrack, "multitrack");
    final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
    assertNotNull(file);
    assertEquals("multitrack", file.getName());
    assertEquals("", file.getDescription());
    assertEquals("10KB", file.getSize());

    final Uri normalizedUri = file.getUri();
    final Uri subUri = normalizedUri.buildUpon().fragment("#2").build();
    try {
      VfsArchive.resolve(subUri);
      unreachable();
    } catch (Exception e) {
      assertNotNull("Thrown exception", e);
      assertEquals("No archive found", e.getMessage());
    }
    {
      final VfsDir browsed = (VfsDir) VfsArchive.browse(file);
      assertTrue(VfsArchive.checkIfArchive(browsed));
      assertEquals(normalizedUri, browsed.getUri());
    }
    {
      final VfsDir browsedCached = (VfsDir) VfsArchive.browseCached(file);
      assertTrue(VfsArchive.checkIfArchive(browsedCached));
      assertEquals(normalizedUri, browsedCached.getUri());
    }
    {
      final VfsDir resolved = (VfsDir) VfsArchive.resolve(uri);
      assertTrue(VfsArchive.checkIfArchive(resolved));
      assertEquals(normalizedUri, resolved.getUri());

      assertEquals(
          "{#1} (RoboCop 3 - Jeroen Tel) <3:56>\n" +
              "{#2} (RoboCop 3 - Jeroen Tel) <0:09>\n" +
              "{#3} (RoboCop 3 - Jeroen Tel) <0:16>\n" +
              "{#4} (RoboCop 3 - Jeroen Tel) <0:24>\n" +
              "{#5} (RoboCop 3 - Jeroen Tel) <0:24>\n" +
              "{#6} (RoboCop 3 - Jeroen Tel) <0:24>\n" +
              "{#7} (RoboCop 3 - Jeroen Tel) <0:12>\n" +
              "{#8} (RoboCop 3 - Jeroen Tel) <0:35>\n" +
              "{#9} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#10} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#11} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#12} (RoboCop 3 - Jeroen Tel) <0:02>\n" +
              "{#13} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#14} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#15} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#16} (RoboCop 3 - Jeroen Tel) <0:05>\n" +
              "{#17} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#18} (RoboCop 3 - Jeroen Tel) <0:01>\n" +
              "{#19} (RoboCop 3 - Jeroen Tel) <0:03>\n" +
              "{#20} (RoboCop 3 - Jeroen Tel) <0:02>\n", listContent(resolved));
    }
    {
      final VfsFile resolvedFile = (VfsFile) VfsArchive.resolve(subUri);
      assertNotNull(resolvedFile);
      assertEquals(subUri, resolvedFile.getUri());
      assertEquals("#2", resolvedFile.getName());
      assertEquals("RoboCop 3 - Jeroen Tel", resolvedFile.getDescription());
      assertEquals("0:09", resolvedFile.getSize());

      // archive dir overrides file
      final VfsDir parentDir = (VfsDir) resolvedFile.getParent();
      assertNotNull(parentDir);
      assertEquals(normalizedUri, parentDir.getUri());
      assertTrue(VfsArchive.checkIfArchive(parentDir));

      final VfsDir parentDirParentDir = (VfsDir) parentDir.getParent();
      assertFalse(VfsArchive.checkIfArchive(parentDirParentDir));
    }
  }

  @Test
  public void testArchive() throws IOException {
    final Uri uri = getFile(app.zxtune.test.R.raw.archive, "archive");
    final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
    assertNotNull(file);
    assertEquals("archive", file.getName());
    assertEquals("", file.getDescription());
    assertEquals("2.4KB", file.getSize());

    final Uri normalizedUri = file.getUri();
    final Uri subFileUri = normalizedUri.buildUpon().fragment("auricom.pt3").build();
    try {
      VfsArchive.resolve(subFileUri);
      unreachable();
    } catch (Exception e) {
      assertNotNull("Thrown exception", e);
      assertEquals("No archive found", e.getMessage());
    }

    final Uri subDirUri = normalizedUri.buildUpon().fragment("coop-Jeffie").build();
    try {
      VfsArchive.resolve(subDirUri);
      unreachable();
    } catch (Exception e) {
      assertNotNull("Thrown exception", e);
      assertEquals("No archive found", e.getMessage());
    }
    {
      final VfsDir browsed = (VfsDir) VfsArchive.browse(file);
      assertTrue(VfsArchive.checkIfArchive(browsed));
      assertEquals(normalizedUri, browsed.getUri());
    }
    {
      final VfsDir browsedCached = (VfsDir) VfsArchive.browseCached(file);
      assertTrue(VfsArchive.checkIfArchive(browsedCached));
      assertEquals(normalizedUri, browsedCached.getUri());
    }
    {
      final VfsDir resolved = (VfsDir) VfsArchive.resolve(uri);
      assertTrue(VfsArchive.checkIfArchive(resolved));
      assertEquals(normalizedUri, resolved.getUri());

      assertEquals(
          "{auricom.pt3} (auricom - sclsmnn^mc ft frolic&fo(?). 2015) <1:13>\n" +
              "[coop-Jeffie] ()\n" +
              " {bass sorrow.pt3} (bass sorrow - scalesmann^mc & jeffie/bw, 2004) <2:10>\n", listContent(resolved));
    }
    {
      final VfsDir resolvedDir = (VfsDir) VfsArchive.resolve(subDirUri);
      assertNotNull(resolvedDir);
      assertEquals(subDirUri, resolvedDir.getUri());
      assertEquals("coop-Jeffie", resolvedDir.getName());
      assertEquals("", resolvedDir.getDescription());
    }
    {
      final VfsFile resolvedFile = (VfsFile) VfsArchive.resolve(subFileUri);
      assertNotNull(resolvedFile);
      assertEquals(subFileUri, resolvedFile.getUri());
      assertEquals("auricom.pt3", resolvedFile.getName());
      assertEquals("auricom - sclsmnn^mc ft frolic&fo(?). 2015", resolvedFile.getDescription());
      assertEquals("1:13", resolvedFile.getSize());

      // archive dir overrides file
      final VfsDir parentDir = (VfsDir) resolvedFile.getParent();
      assertNotNull(parentDir);
      assertEquals(normalizedUri, parentDir.getUri());
      assertTrue(VfsArchive.checkIfArchive(parentDir));

      final VfsDir parentDirParentDir = (VfsDir) parentDir.getParent();
      assertFalse(VfsArchive.checkIfArchive(parentDirParentDir));
    }
  }

  @Test
  public void testForceResolve() throws IOException {
    {
      final Uri uri = getFile(app.zxtune.test.R.raw.gzipped, "gzipped");
      final VfsFile file = (VfsFile) VfsArchive.resolve(uri);

      final Uri normalizedUri = file.getUri();
      final Uri subUri = normalizedUri.buildUpon().fragment("+unGZIP").build();
      {
        final VfsFile resolvedFile = (VfsFile) VfsArchive.resolveForced(subUri,
            StubProgressCallback.instance());
        assertNotNull(resolvedFile);
        assertEquals(subUri, resolvedFile.getUri());
        assertEquals("+unGZIP", resolvedFile.getName());
        assertEquals("sll3", resolvedFile.getDescription());
        assertEquals("2:33", resolvedFile.getSize());

        // no parent dir really, so just file itself
        final VfsFile resolvedFileParent = (VfsFile) resolvedFile.getParent();
        assertNotNull(resolvedFileParent);
        assertEquals(normalizedUri, resolvedFileParent.getUri());
      }
    }
    {
      final Uri uri = getFile(app.zxtune.test.R.raw.multitrack, "multitrack");
      final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
      assertNotNull(file);

      final Uri normalizedUri = file.getUri();
      final Uri subUri = normalizedUri.buildUpon().fragment("#2").build();
      {
        final VfsFile resolvedFile = (VfsFile) VfsArchive.resolveForced(subUri,
            StubProgressCallback.instance());
        assertNotNull(resolvedFile);
        assertEquals(subUri, resolvedFile.getUri());
        assertEquals("#2", resolvedFile.getName());
        assertEquals("RoboCop 3 - Jeroen Tel", resolvedFile.getDescription());
        assertEquals("0:09", resolvedFile.getSize());

        // archive dir overrides file
        final VfsDir parentDir = (VfsDir) resolvedFile.getParent();
        assertNotNull(parentDir);
        assertEquals(normalizedUri, parentDir.getUri());
        assertTrue(VfsArchive.checkIfArchive(parentDir));

        final VfsDir parentDirParentDir = (VfsDir) parentDir.getParent();
        assertFalse(VfsArchive.checkIfArchive(parentDirParentDir));
      }
    }
    {
      final Uri uri = getFile(app.zxtune.test.R.raw.archive, "archive");
      final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
      assertNotNull(file);

      final Uri normalizedUri = file.getUri();
      final Uri subFileUri = normalizedUri.buildUpon().fragment("auricom.pt3").build();
      {
        final VfsFile resolvedFile = (VfsFile) VfsArchive.resolveForced(subFileUri,
            StubProgressCallback.instance());
        assertNotNull(resolvedFile);
        assertEquals(subFileUri, resolvedFile.getUri());
        assertEquals("auricom.pt3", resolvedFile.getName());
        assertEquals("auricom - sclsmnn^mc ft frolic&fo(?). 2015", resolvedFile.getDescription());
        assertEquals("1:13", resolvedFile.getSize());

        // archive dir overrides file
        final VfsDir parentDir = (VfsDir) resolvedFile.getParent();
        assertNotNull(parentDir);
        assertEquals(normalizedUri, parentDir.getUri());
        assertTrue(VfsArchive.checkIfArchive(parentDir));

        final VfsDir parentDirParentDir = (VfsDir) parentDir.getParent();
        assertFalse(VfsArchive.checkIfArchive(parentDirParentDir));
      }
    }
    {
      final Uri uri = getFile(app.zxtune.test.R.raw.archive, "archive2");
      final VfsFile file = (VfsFile) VfsArchive.resolve(uri);
      assertNotNull(file);

      final Uri normalizedUri = file.getUri();
      final Uri subDirUri = normalizedUri.buildUpon().fragment("coop-Jeffie").build();
      {
        final VfsDir resolvedDir = (VfsDir) VfsArchive.resolveForced(subDirUri,
            StubProgressCallback.instance());
        assertNotNull(resolvedDir);
        assertEquals(subDirUri, resolvedDir.getUri());
        assertEquals("coop-Jeffie", resolvedDir.getName());
        assertEquals("", resolvedDir.getDescription());
      }
    }
  }

  private static String listContent(VfsDir dir) {
    final StringBuilder list = new StringBuilder();
    listDir(dir, list, "");
    return list.toString();
  }

  private static void listDir(VfsDir input, final StringBuilder list, final String tab) {
    try {
      input.enumerate(new VfsDir.Visitor() {
        @Override
        public void onItemsCount(int count) {
        }

        @Override
        public void onDir(VfsDir dir) {
          list.append(String.format("%s[%s] (%s)\n", tab, dir.getName(), dir.getDescription()));
          listDir(dir, list, tab + " ");
        }

        @Override
        public void onFile(VfsFile file) {
          list.append(String.format("%s{%s} (%s) <%s>\n", tab, file.getName(), file.getDescription(), file.getSize()));
        }
      });
    } catch (Exception e) {
      list.append(e);
    }
  }
}

File: apps/zxtune-android/zxtune/src/androidTest/java/app/zxtune/fs/VfsArchiveTest.kt
package app.zxtune.fs

import android.content.res.Resources
import android.net.Uri
import androidx.annotation.RawRes
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import app.zxtune.assertThrows
import app.zxtune.io.Io.copy
import app.zxtune.io.TransactionalOutputStream
import app.zxtune.test.R
import app.zxtune.utils.StubProgressCallback
import org.junit.After
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.BufferedInputStream
import java.io.File
import java.io.IOException

@RunWith(AndroidJUnit4::class)
class VfsArchiveTest {

    private lateinit var resources: Resources
    private lateinit var tmpDir: File

    private fun getFile(@RawRes res: Int, filename: String) = with(File(tmpDir, filename)) {
        BufferedInputStream(resources.openRawResource(res)).use { input ->
            TransactionalOutputStream(this).use {
                val size = copy(input, it)
                assertTrue(size > 0)
                it.flush()
            }
        }
        Uri.Builder().scheme("file").path(path).build()
    }

    @Before
    fun setUp() {
        resources = InstrumentationRegistry.getInstrumentation().context.resources
        tmpDir = File(
            System.getProperty("java.io.tmpdir", "."),
            "VfsArchiveTest/${System.currentTimeMillis()}"
        )
    }

    @After
    fun tearDown() {
        tmpDir.listFiles()?.map(File::delete)
        tmpDir.delete()
    }

    @Test
    fun testNull() {
        assertThrows<NullPointerException> {
            VfsArchive.resolve(null)
        }
        assertThrows<NullPointerException> {
            VfsArchive.browse(null)
        }
        assertThrows<NullPointerException> {
            VfsArchive.browseCached(null)
        }
    }

    @Test
    fun testDir() = Uri.parse("file:/proc").let { id ->
        (VfsArchive.resolve(id) as VfsDir).run {
            assertFalse(VfsArchive.checkIfArchive(this))
            assertEquals("proc", name)
            assertEquals("", description)
            assertEquals(id, uri)
        }
    }

    @Test
    fun testFile() {
        assertResolvedAsFile(Uri.parse("file:/proc/meminfo"), expectedSize = "0")
    }

    @Test
    fun testFileTrack() = getFile(R.raw.track, "track").let { id ->
        val file = assertResolvedAsFile(id, expectedSize = "2.8K")
        assertSame(file, VfsArchive.browse(file))
        assertSame(file, VfsArchive.browseCached(file))
        assertEquals(file.uri, VfsArchive.resolve(id)?.uri)
    }

    @Test
    fun testGzippedTrack() = getFile(R.raw.gzipped, "gzipped").let { id ->
        val file = assertResolvedAsFile(id, expectedSize = "54.1K")
        val subUri = file.uri.withFragment("+unGZIP")
        assertThrows<IOException> {
            VfsArchive.resolve(subUri)
        }.let { e ->
            assertEquals("No archive found", e.message)
        }
        assertEquals(file, VfsArchive.browse(file))
        assertEquals(file, VfsArchive.browseCached(file))
        assertEquals(file.uri, VfsArchive.resolve(id)?.uri)
        assertResolvedAsFile(
            subUri, expectedName = "gzipped", expectedDescription = "sll3", expectedSize = "2:33"
        ).run {
            // no parent dir really, so just file itself
            assertEquals(file.uri, (parent as VfsFile).uri)
        }
    }

    @Test
    fun testMultitrack() = getFile(R.raw.multitrack, "multitrack").let { id ->
        val file = assertResolvedAsFile(id, expectedSize = "16.2K")
        val subUri = file.uri.withFragment("#2")
        assertThrows<IOException> {
            VfsArchive.resolve(subUri)
        }.let { e ->
            assertEquals("No archive found", e.message)
        }
        (VfsArchive.browse(file) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
        }
        (VfsArchive.browseCached(file) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
        }
        (VfsArchive.resolve(id) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
            assertArrayEquals(
                arrayOf(
                    File(0, "#1", "#1", "Title - Tim and/or Geoff Follin", "0:16"),
                    File(0, "#2", "#2", "Looney Tunes - Tim and/or Geoff Follin", "3:12"),
                    File(0, "#3", "#3", "Like Cats And Mice - Tim and/or Geoff Follin", "3:16"),
                    File(0, "#4", "#4", "Victory is Mine! - Tim and/or Geoff Follin", "0:10"),
                    File(0, "#5", "#5", "Game Over - Tim and/or Geoff Follin", "0:10"),
                ), listContent(this)
            )
        }
        assertResolvedAsFile(
            subUri,
            expectedName = "#2",
            expectedDescription = "Looney Tunes - Tim and/or Geoff Follin",
            expectedSize = "3:12"
        ).run {
            // archive dir overrides file
            (parent as VfsDir).run {
                assertEquals(file.uri, uri)
                assertTrue(VfsArchive.checkIfArchive(this))
                assertFalse(VfsArchive.checkIfArchive(parent as VfsDir))
            }
        }
    }

    @Test
    fun testArchive() = getFile(R.raw.archive, "archive").let { id ->
        val file = assertResolvedAsFile(id, expectedSize = "2.4K")
        val subFileUri = file.uri.withFragment("auricom.pt3")
        assertThrows<IOException> {
            VfsArchive.resolve(subFileUri)
        }.let { e ->
            assertEquals("No archive found", e.message)
        }
        val subDirUri = file.uri.withFragment("coop-Jeffie")
        assertThrows<IOException> {
            VfsArchive.resolve(subDirUri)
        }.let { e ->
            assertEquals("No archive found", e.message)
        }
        (VfsArchive.browse(file) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
        }
        (VfsArchive.browseCached(file) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
        }
        (VfsArchive.resolve(id) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
            assertArrayEquals(
                arrayOf(
                    File(
                        0,
                        "auricom.pt3",
                        "auricom.pt3",
                        "auricom - sclsmnn^mc ft frolic&fo(?). 2015",
                        "1:13"
                    ),
                    Dir(0, "coop-Jeffie", "coop-Jeffie"),
                    File(
                        1,
                        "coop-Jeffie/bass sorrow.pt3",
                        "bass sorrow.pt3",
                        "bass sorrow - scalesmann^mc & jeffie/bw, 2004",
                        "2:10"
                    ),
                ), listContent(this)
            )
        }
        (VfsArchive.resolve(subDirUri) as VfsDir).run {
            assertEquals(subDirUri, uri)
            assertEquals("coop-Jeffie", name)
            assertEquals("", description)
            (parent as VfsDir).run {
                assertEquals(file.uri, uri)
                assertTrue(VfsArchive.checkIfArchive(this))
                assertFalse(VfsArchive.checkIfArchive(parent as VfsDir))
            }
            assertArrayEquals(
                arrayOf(
                    File(
                        0,
                        "coop-Jeffie/bass sorrow.pt3",
                        "bass sorrow.pt3",
                        "bass sorrow - scalesmann^mc & jeffie/bw, 2004",
                        "2:10"
                    ),
                ), listContent(this)
            )
        }
        assertResolvedAsFile(
            subFileUri,
            expectedName = "auricom.pt3",
            expectedDescription = "auricom - sclsmnn^mc ft frolic&fo(?). 2015",
            expectedSize = "1:13"
        ).run {
            // archive dir overrides file
            (parent as VfsDir).run {
                assertEquals(file.uri, uri)
                assertTrue(VfsArchive.checkIfArchive(this))
                assertFalse(VfsArchive.checkIfArchive(parent as VfsDir))
            }
        }
    }

    @Test
    fun testArchiveWithGzips() = getFile(R.raw.archive_with_gzips, "archive_with_gzips").let { id ->
        val file = assertResolvedAsFile(id, expectedSize = "67.2K")
        (VfsArchive.browse(file) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
        }
        (VfsArchive.browseCached(file) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
        }
        (VfsArchive.resolve(id) as VfsDir).run {
            assertTrue(VfsArchive.checkIfArchive(this))
            assertEquals(file.uri, uri)
            assertArrayEquals(
                arrayOf(
                    File(0, "gzipped/+unGZIP", "gzipped", "sll3", "2:33"),
                    Dir(0, "gzipped_archive/+unGZIP", "gzipped_archive"),
                    File(
                        1,
                        "gzipped_archive/+unGZIP/auricom.pt3",
                        "auricom.pt3",
                        "auricom - sclsmnn^mc ft frolic&fo(?). 2015",
                        "1:13"
                    ),
                    Dir(1, "gzipped_archive/+unGZIP/coop-Jeffie", "coop-Jeffie"),
                    File(
                        2,
                        "gzipped_archive/+unGZIP/coop-Jeffie/bass sorrow.pt3",
                        "bass sorrow.pt3",
                        "bass sorrow - scalesmann^mc & jeffie/bw, 2004",
                        "2:10"
                    ),
                    Dir(0, "gzipped_multitrack/+unGZIP", "gzipped_multitrack"),
                    File(
                        1,
                        "gzipped_multitrack/+unGZIP/#1",
                        "#1",
                        "Title - Tim and/or Geoff Follin",
                        "0:16"
                    ),
                    File(
                        1,
                        "gzipped_multitrack/+unGZIP/#2",
                        "#2",
                        "Looney Tunes - Tim and/or " + "Geoff Follin",
                        "3:12"
                    ),
                    File(
                        1,
                        "gzipped_multitrack/+unGZIP/#3",
                        "#3",
                        "Like Cats And Mice - Tim " + "and/or Geoff Follin",
                        "3:16"
                    ),
                    File(
                        1,
                        "gzipped_multitrack/+unGZIP/#4",
                        "#4",
                        "Victory is Mine! - Tim" + " and/or Geoff Follin",
                        "0:10"
                    ),
                    File(
                        1,
                        "gzipped_multitrack/+unGZIP/#5",
                        "#5",
                        "Game Over - Tim " + "and/or Geoff Follin",
                        "0:10"
                    ),
                    File(
                        0,
                        "track",
                        "track",
                        "AsSuRed ... Hi! My Frends ... - Mm<M of Sage 14.Apr.XX twr 00:37",
                        "3:41"
                    ),
                ), listContent(this)
            )
            assertFalse(VfsArchive.checkIfArchive(parent as VfsDir))
        }
        val gzippedArchiveId = id.withFragment("gzipped_archive")
        assertThrows<IOException> {
            VfsArchive.resolve(gzippedArchiveId)
        }
        val gzippedArchiveSubdirId = id.withFragment("gzipped_archive/+unGZIP")
        (VfsArchive.resolve(gzippedArchiveSubdirId) as VfsDir).run {
            assertEquals("gzipped_archive", name)
            assertEquals(gzippedArchiveSubdirId, uri)
            (parent as VfsDir).run {
                assertEquals(file.uri, uri)
            }
        }
        val gzippedMultitrackId = id.withFragment("gzipped_multitrack")
        assertThrows<IOException> {
            VfsArchive.resolve(gzippedMultitrackId)
        }
        val gzippedMultitrackSubdirId = id.withFragment("gzipped_multitrack/+unGZIP")
        (VfsArchive.resolve(gzippedMultitrackSubdirId) as VfsDir).run {
            assertEquals("gzipped_multitrack", name)
            assertEquals(gzippedMultitrackSubdirId, uri)
            (parent as VfsDir).run {
                assertEquals(file.uri, uri)
            }
        }
    }

    @Test
    fun testForceResolve() {
        getFile(R.raw.gzipped, "gzipped").let { id ->
            val file = assertResolvedAsFile(id, expectedSize = "54.1K")
            val subUri = file.uri.withFragment("+unGZIP")
            (VfsArchive.resolveForced(subUri, StubProgressCallback.instance()) as VfsFile).run {
                assertEquals(subUri, uri)
                assertEquals("gzipped", name)
                assertEquals("sll3", description)
                assertEquals("2:33", size)

                // no parent dir really, so just file itself
                assertEquals(file.uri, (parent as VfsFile).uri)
            }
        }
        getFile(R.raw.multitrack, "multitrack").let { id ->
            val file = assertResolvedAsFile(id, expectedSize = "16.2K")
            val subUri = file.uri.withFragment("#2")
            (VfsArchive.resolveForced(subUri, StubProgressCallback.instance()) as VfsFile).run {
                assertEquals(subUri, uri)
                assertEquals("#2", name)
                assertEquals("Looney Tunes - Tim and/or Geoff Follin", description)
                assertEquals("3:12", size)

                // archive dir overrides file
                (parent as VfsDir).run {
                    assertEquals(file.uri, uri)
                    assertTrue(VfsArchive.checkIfArchive(this))
                    (parent as VfsDir).run {
                        assertFalse(VfsArchive.checkIfArchive(this))
                    }
                }
            }
        }

        getFile(R.raw.archive, "archive").let { id ->
            val file = assertResolvedAsFile(id, expectedSize = "2.4K")
            val subFileUri = file.uri.withFragment("auricom.pt3")

            (VfsArchive.resolveForced(
                subFileUri, StubProgressCallback.instance()
            ) as VfsFile).run {
                assertEquals(subFileUri, uri)
                assertEquals("auricom.pt3", name)
                assertEquals("auricom - sclsmnn^mc ft frolic&fo(?). 2015", description)
                assertEquals("1:13", size)

                // archive dir overrides file
                (parent as VfsDir).run {
                    assertEquals(file.uri, uri)
                    assertTrue(VfsArchive.checkIfArchive(this))
                    (parent as VfsDir).run {
                        assertFalse(VfsArchive.checkIfArchive(this))
                    }
                }
            }
        }
        getFile(R.raw.archive, "archive2").let { id ->
            val file = assertResolvedAsFile(id, expectedSize = "2.4K")
            val subDirUri = file.uri.withFragment("coop-Jeffie")
            (VfsArchive.resolveForced(
                subDirUri, StubProgressCallback.instance()
            ) as VfsDir).run {
                assertEquals(subDirUri, uri)
                assertEquals("coop-Jeffie", name)
                assertEquals("", description)
            }
        }
    }
}

private fun assertResolvedAsFile(
    id: Uri,
    expectedName: String = requireNotNull(id.lastPathSegment),
    expectedDescription: String = "",
    expectedSize: String
) = (VfsArchive.resolve(id) as VfsFile).apply {
    assertEquals(id, uri)
    assertEquals(expectedName, name)
    assertEquals(expectedDescription, description)
    assertEquals(expectedSize, size)
}

fun Uri.withFragment(fragment: String): Uri = buildUpon().fragment(fragment).build()

sealed class Entry

data class File(
    val depth: Int, val subpath: String, val name: String, val description: String, val size: String
) : Entry()

data class Dir(val depth: Int, val subpath: String, val name: String) : Entry()

private fun listContent(dir: VfsDir) = ArrayList<Entry>().apply {
    listDir(dir, this, 0)
}.toArray()

private fun listDir(input: VfsDir, list: ArrayList<Entry>, depth: Int) {
    input.enumerate(object : VfsDir.Visitor() {
        override fun onDir(dir: VfsDir) = with(dir) {
            require(description.isEmpty())
            list.add(Dir(depth, requireNotNull(uri.fragment), name))
            listDir(dir, list, depth + 1)
        }

        override fun onFile(file: VfsFile) = with(file) {
            list.add(File(depth, requireNotNull(uri.fragment), name, description, size))
            Unit
        }
    })
}
